### DISCUSSION TITLE
HttpClient causes exception in Droid app with Let's Encrypt certificates

### DISCUSSION THREAD

**Discussion by luciusinfabula** [unoplatform/uno#9884] [Q&A]
### Current behavior

Dears,
I'm using this library [FeedReader](https://github.com/arminreiter/FeedReader/) to create a Rss Reader.
The library is based on HttpClient library.
Trying to retrieve data from a site protected by a Let's Encrypt TLS certificate causes following exception:

> {System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception. ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception. ---> Mono.Btls.MonoBtlsException: Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
  at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/external/boringssl/ssl/handshake_client.c:1132

In WPF/UWP project all goes ok while in DROID project this behaviour happens only for Let's Encrypt sites.

I tried 5 sites with Let's Encrypt and all return the same exception, while 5 ones with different TLS certificates no exception is raised.

List of Let's Encrypt sites:
- https://it.quora.com
- https://letsencrypt.org/about
- https://rbv.biblioteche.it/events/rss?start=0

List of sites with different certificates:
- https://www.chess.com
- https://platform.uno
- https://microsoft.com

Luc

### Affected platforms

Android

### IDE

Visual Studio 2019

### IDE version

16.11.16

**Comment by MartinZikmund** [comment]
Can you try one of the options here https://learn.microsoft.com/en-us/xamarin/android/app-fundamentals/http-stack?tabs=windows to see if some of them helps?

Alternative is to ignore the certificate validation altogether, but that is of course not ideal:

```
httpClientHandler.ServerCertificateCustomValidationCallback +=
    (sender, cert, chain, sslPolicyErrors) => true;
```

<blockquote><img src="https://learn.microsoft.com/en-us/media/logos/logo-ms-social.png" width="48" align="right"><div><strong><a href="https://learn.microsoft.com/en-us/xamarin/android/app-fundamentals/http-stack">HttpClient Stack and SSL/TLS Implementation Selector for Android - Xamarin</a></strong></div><div>The HttpClient Stack and SSL/TLS Implementation selectors determine the HttpClient and SSL/TLS implementation that will be used by your Xamarin.Android apps.</div></blockquote>

**Comment by luciusinfabula** [comment]
Hi Martin,
thanks for your kindly reply.

I tried successfuly both changing HttpClient handling in advanced Droid property configuration and using explicit AndroidClientHandler and they work perfect.

At the end I decided to implement the second solution using Platform-specific C# directive like following [click here](https://pastebin.com/eai0wUew) .

And now I can handle in Droid project https sites configured with Let's Encrypt certificates too.

Many thanks again for your help.
Luc
